src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -790,6 +804,37 @@ internal static string UnmangleMetadataNameForArity(string emittedTypeName, int
             } // using
 
             types = nestedTypes;
+
+            // Merge pairs with the same key
+            if (possiblyHavePairsWithDuplicateKey)
+            {
+                var names = new Dictionary<string, int>(nameComparer);
```

Comment:
Does it make sense to set the capacity here to `nestedNamespaces.Count` given that is at the least an upper bound on the number of entries.